jsonpointer.py

```diff
@@ -29,40 +29,50 @@
 # (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 # THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #
+from __future__ import annotations
 
 """ Identify specific nodes in a JSON document (RFC 6901) """
 
-from __future__ import unicode_literals
-
 # Will be parsed by setup.py to determine package metadata
-__author__ = 'Stefan Kögl <stefan@skoegl.net>'
-__version__ = '2.3'
-__website__ = 'https://github.com/stefankoegl/python-json-pointer'
-__license__ = 'Modified BSD License'
+__author__ = "Stefan Kögl <stefan@skoegl.net>"
+__version__ = "3.0"
+__website__ = "https://github.com/stefankoegl/python-json-pointer"
+__license__ = "Modified BSD License"
 
 
-try:
-    from itertools import izip
-    str = unicode
-    encode_str = lambda u: u.encode("raw_unicode_escape")
-except ImportError:  # Python 3
-    izip = zip
-    encode_str = lambda u: u
+import copy
+import re
+from collections.abc import Mapping, Sequence
+from itertools import tee, chain
+from typing import *
 
-try:
-    from collections.abc import Mapping, Sequence
-except ImportError:  # Python 3
-    from collections import Mapping, Sequence
+T = TypeVar("T")
+TJsonPointer = TypeVar("TJsonPointer", bound="JsonPointer")
+JsonType = Union[Dict[str, "JsonType"], List["JsonType"], str, bool, int, float, None]
+_Parts = Union[str, int]
+_Suffixes = Union["JsonPointer", str, Iterable[str]]
+_Undefined = Union[T, "_Nothing"]
 
-from itertools import tee, chain
-import re
-import copy
 
+class _Singleton(type):
```

Comment:
This may be overkill, but it made typing the `Nothing` object hints way easier to have a class it's an instance of. It could have been an empty `class _Nothing: pass`, **but** with ~5 lines more of code it gets the really nice property that every instance of `_Nothing` is the same instance.

Without that, checking `default is _nothing` would fail, as `is` checks `id(lhs) == id(rhs)` (aka "are they the same instance"? You can work around that by instead checking `isinstance(x, _Nothing)`, but I prefer the syntax you had of `is _nothing`